src/hapi/pipelines/utilities/error_handling.py

```diff
@@ -0,0 +1,175 @@
+            message_type (str): The type of message (error or warning). Default is "error"
+            flag_to_hdx (bool): Flag indicating if the message should be added to HDX metadata. Default is False
+        Returns:
+            None
+        """
+        text = f"{value_type} {value} not found"
+        self.add_message(
+            pipeline,
+            identifier,
+            text,
+            resource_name,
+            message_type,
+            flag_to_hdx,
+        )
+
+    def add_multi_valued_message(
+        self,
+        pipeline: str,
+        identifier: str,
+        text: str,
+        values: List,
+        resource_name: str = None,
+        message_type: str = "error",
+        flag_to_hdx: bool = False,
+    ) -> bool:
+        """
+        Add a new message (typically a warning or error) concerning a list of
+        values to a set of messages in a fixed format:
+            pipeline - identifier - n {text}. First 10 values: n1,n2,n3...
+        If less than 10 values, ". First 10 values" is omitted. identifier is usually
+        a dataset name.
+        Args:
+            pipeline (str): Name of the scaper originating the error
+            identifier (str): Identifier e.g. dataset name
+            text (str): Text to use e.g. "negative values removed"
+            values (List[str]): List of values of concern
+            resource_name (str): The resource name that the message applies to. Only needed if flagging on HDX
+            message_type (str): The type of message (error or warning). Default is "error"
+            flag_to_hdx (bool): Flag indicating if the message should be added to HDX metadata. Default is False
+        Returns:
+            bool: True if a message was added, False if not
+        """
+        if not values:
+            return False
+        no_values = len(values)
+        if no_values > 10:
+            values = values[:10]
+            msg = ". First 10 values"
+        else:
+            msg = ""
+        text = f"{no_values} {text}{msg}: {', '.join(values)}"
+        self.add_message(
+            pipeline,
+            identifier,
+            text,
+            resource_name,
+            message_type,
+            flag_to_hdx,
+        )
+        return True
+
+    def output_errors(self, flag_to_hdx: bool) -> None:
+        for _, errors in self.shared_errors["error"].items():
+            errors = sorted(errors)
+            for error in errors:
+                logger.error(error)
+        for _, warnings in self.shared_errors["warning"].items():
+            warnings = sorted(warnings)
+            for warning in warnings:
+                logger.warning(warning)
+        if flag_to_hdx:
+            for identifier, errors in self.shared_errors["hdx_error"].items():
+                write_error_to_resource(identifier, errors)
+
+
+def write_error_to_resource(identifier: Tuple[str], errors: set[str]) -> bool:
```

Comment:
I think this should be in HDX Python API. As we move pipeline logic out of HAPI into scrapers I guess we'll need to call this from there.

Reply:
That makes sense - I can add it before we merge this.